src/core_scan/Core_scan.ml

```diff
@@ -854,15 +842,17 @@ let scan ?match_hook config ((rules, invalid_rules), rules_parse_time) :
                     | `Steps _ ->
                         true)
              |> List.filter (fun r ->
-                    (* TODO: some of this is already done in pysemgrep, so maybe
-                     * we should guard with a flag that only osemgrep set
-                     * like Core_scan_config.paths_processing: bool?
-                     *)
+                    (* Honor per-rule include/exclude *)
                     match r.R.paths with
                     | None -> true
                     | Some paths -> Filter_target.filter_paths paths file)
            in
-           let xtarget = xtarget_of_file config xlang file in
+           let was_scanned =
+             match applicable_rules with
+             | [] -> false
+             | _ -> true
+           in
```

Comment:
If there aren't applicable rules, are we able to exit this function faster? I'm asking partly to avoid doing work we don't have to, partly because this feels a little fragile

Reply:
We should be able to skip the rest if there aren't any applicable rules. We should do this without raising exceptions and by breaking this giant function into smaller ones.

Reply:
It's too complicated for this PR due to progress hooks and the whole thing happening in a child process. I'm leaving notes about the need for clarification for this and that.